src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.RuntimeIdentifierInference.targets

```
@@ -61,6 +61,42 @@ Copyright (c) .NET Foundation. All rights reserved.
     <RuntimeIdentifier Condition="'$(PlatformTarget)' == 'x86' or '$(PlatformTarget)' == ''">win7-x86</RuntimeIdentifier>
   </PropertyGroup>
 
+  <!-- Check if SelfContained is global, resets the value to it's original state if it was local.-->
+  <PropertyGroup>
+    <_OriginalSelfContainedValue>$(SelfContained)</_OriginalSelfContainedValue>
+    <SelfContained></SelfContained>
+  </PropertyGroup>
+
+  <Choose>
+    <!-- Assumes a global variable cannot be an undefined variable.-->
+    <When Condition="'$(SelfContained)' != '' and $(SelfContained) == $(_OriginalSelfContainedValue)">
+      <PropertyGroup>
+        <_SelfContainedIsGlobal>true</_SelfContainedIsGlobal>
+      </PropertyGroup>
+    </When>
+    <Otherwise>
+      <PropertyGroup>
+        <_SelfContainedIsGlobal>false</_SelfContainedIsGlobal>
+      </PropertyGroup>
+    </Otherwise>
+  </Choose>
+
+  <PropertyGroup>
+    <SelfContained>$(_OriginalSelfContainedValue)</SelfContained>
+    <_OriginalSelfContainedValue></_OriginalSelfContainedValue>
+  </PropertyGroup>
+
+  <!-- PubishSelfContained may be set but not change self contained if we are not publishing, or self contained was already set globally. This condition tracks whether it will take effect.-->
+  <PropertyGroup Condition=" '$(_CommandLineDefinedSelfContained)' != 'true' and '$(_SelfContainedIsGlobal)' != 'true' and '$(_IsPublishing)' == 'true' and ('$(PublishSelfContained)' == 'true' or '$(PublishSelfContained)' == 'false')">
+    <_PublishSelfContainedIsActive>true</_PublishSelfContainedIsActive>
+  </PropertyGroup>
+
+  <!-- Edit SelfContained to match the value of PublishSelfContained if we are publishing.
+       This Won't affect t:/Publish (because of _IsPublishing), and also won't override a global SelfContained property.-->
+  <PropertyGroup Condition="'$(_PublishSelfContainedIsActive)' == 'true'">
+    <SelfContained>$(PublishSelfContained)</SelfContained>
+  </PropertyGroup>
```

Do we need `_PublishSelfContainedIsActive` and whether `SelfContained` was specified in a global property at all?  Could we do the following instead?

```suggestion
  <PropertyGroup Condition="'$(PublishSelfContained)' != ''">
    <SelfContained>$(PublishSelfContained)</SelfContained>
  </PropertyGroup>
```

This would have no effect if `SelfContained` was a global property, but we wouldn't need all that extra logic to check.

---

I don't think we can do that exact proposal, we still need to check `_IsPublishing`. And I don't think we can just check `!= ''` because it would fail if someone passed a value besides `true` or `false`. 

But overall, I agree we can slim this down a lot and achieve the same result. I tried to do this somewhere else to decouple SC from RidInference but ended up needing to move it here anyways and some stuff got left.

---

I think the valid values for `PublishSelfContained` are the same as the valid values for `SelfContained`: basically True, False, and empty.  So I'm not sure if we need to worry about whether or not we copy an invalid value.

---

If we do the check as true or false, it will ignore PublishSelfContained silently. If we do the check with `!= ''` it produces the following:
```go
MSBuild version 17.4.0+18d5aef85 for .NET
  Determining projects to restore...
C:\Users\\sdk\artifacts\bin\redist\Debug\dotnet\sdk\7.0.100-dev\Sdks\Microsoft.
NET.Sdk\targets\Microsoft.NET.Sdk.FrameworkReferenceResolution.targets(99,33): error MSB4
030: "Blah" is an invalid value for the "SelfContained" parameter of the "ProcessFrameworkR
eferences" task. The "SelfContained" parameter is of type "System.Boolean". [C:\Users\noa
hgilson\sdk\newc\newc.csproj]****
```

So I'll add a new error for `PublishSelfContained`. Thank you for your help, I made this way too complicated